Dockerfile

```diff
@@ -278,42 +278,75 @@ RUN echo "Pip version: ${PIP_VERSION}"
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/setup.cfg >${AIRFLOW_SOURCES}/setup.cfg \
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/version.py >${AIRFLOW_SOURCES}/airflow/version.py \
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/__init__.py >${AIRFLOW_SOURCES}/airflow/__init__.py \
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/bin/airflow >${AIRFLOW_SOURCES}/airflow/bin/airflow
+
+# Airflow Extras installed
+ARG AIRFLOW_EXTRAS="all"
+ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
+RUN echo "Installing with extras: ${AIRFLOW_EXTRAS}."
+
+RUN pip install --no-use-pep517 -e ".[${AIRFLOW_EXTRAS}]"
+
+# Note! We are copying everything with airflow:airflow user:group even if we use root to run the scripts
 # This is fine as root user will be able to use those dirs anyway.
 
 # Airflow sources change frequently but dependency configuration won't change that often
 # We copy setup.py and other files needed to perform setup of dependencies
-# This way cache here will only be invalidated if any of the
-# version/setup configuration change but not when airflow sources change
+# So in case setup.py changes we can install latest dependencies required.
 COPY --chown=airflow:airflow setup.py ${AIRFLOW_SOURCES}/setup.py
 COPY --chown=airflow:airflow setup.cfg ${AIRFLOW_SOURCES}/setup.cfg
 
 COPY --chown=airflow:airflow airflow/version.py ${AIRFLOW_SOURCES}/airflow/version.py
 COPY --chown=airflow:airflow airflow/__init__.py ${AIRFLOW_SOURCES}/airflow/__init__.py
 COPY --chown=airflow:airflow airflow/bin/airflow ${AIRFLOW_SOURCES}/airflow/bin/airflow
 
-# Airflow Extras installed
-ARG AIRFLOW_EXTRAS="all"
-ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
-RUN echo "Installing with extras: ${AIRFLOW_EXTRAS}."
-
-# First install only dependencies but no Apache Airflow itself
-# This way regular changes in sources of Airflow will not trigger reinstallation of all dependencies
-# And this Docker layer will be reused between builds.
+# The goal of this line is to install the dependencies from the most current setup.py from sources
+# This will be usually incremental small set of packages so it will be very fast
 RUN pip install --no-use-pep517 -e ".[${AIRFLOW_EXTRAS}]"
 
 COPY --chown=airflow:airflow airflow/www/package.json ${AIRFLOW_SOURCES}/airflow/www/package.json
 COPY --chown=airflow:airflow airflow/www/package-lock.json ${AIRFLOW_SOURCES}/airflow/www/package-lock.json
 
 WORKDIR ${AIRFLOW_SOURCES}/airflow/www
 
+ARG BUILD_NPM=true
+ENV BUILD_NPM=${BUILD_NPM}
+
 # Install necessary NPM dependencies (triggered by changes in package-lock.json)
-RUN gosu ${AIRFLOW_USER} npm ci
+RUN \
+    if [[ "${BUILD_NPM}" == "true" ]]; then \
+        gosu ${AIRFLOW_USER} npm ci; \
+    fi
 
 COPY --chown=airflow:airflow airflow/www/ ${AIRFLOW_SOURCES}/airflow/www/
 
 # Package NPM for production
-RUN gosu ${AIRFLOW_USER} npm run prod
+RUN \
+    if [[ "${BUILD_NPM}" == "true" ]]; then \
+        gosu ${AIRFLOW_USER} npm run prod; \
+    fi
```

Comment:
Probably a further enhancment (i.e. not in this PR) but is it worth making a separate npm-only image and doing the `npm run prod` in there.

Maybe this is where the flavours of image come in - for ci/dev having Npm in the same image makes sense, but for prod it would be nice to not have to install node.

Reply:
Absolutely. I will include that as part of proposal for Production image discussion!